### PR TITLE
Add mocked_host library

### PR DESCRIPTION
**Pull request by chfast**
It takes the MockedHost implementation from evmone and creates an interface library out of it.

Part of #447.

### CODE REVIEW

File: include/evmc/mocked_host.hpp [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+                                    const bytes32& value) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return EVMC_STORAGE_UNCHANGED;
+
+        auto& old = it->second.storage[key];
+
+        // Follow https://eips.ethereum.org/EIPS/eip-1283 specification.
```
Comment (axic) [inline include/evmc/mocked_host.hpp]: 2200 ?

Comment (chfast) [reply include/evmc/mocked_host.hpp]: No.

File: include/evmc/mocked_host.hpp [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+        /// The log topics.
+        std::vector<bytes32> topics;
+
+        /// Equal operator.
+        bool operator==(const log_record& other) const noexcept
+        {
+            return creator == other.creator && data == other.data &&
+                   topics.size() == other.topics.size() &&
+                   std::equal(topics.begin(), topics.end(), other.topics.begin());
```
Comment (gumb0) [inline include/evmc/mocked_host.hpp]: wouldn't `topics == other.topics` work here?

File: include/evmc/mocked_host.hpp [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        auto& old = it->second.storage[key];
+
+        // Follow https://eips.ethereum.org/EIPS/eip-1283 specification.
+        // WARNING! This is not complete implementation as refund is not handled here.
+
+        if (old.value == value)
+            return EVMC_STORAGE_UNCHANGED;
+
+        evmc_storage_status status;
+        {
```
Comment (gumb0) [inline include/evmc/mocked_host.hpp]: what's the point of this block?

File: include/evmc/mocked_host.hpp [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        if (n > 0)
+            std::copy_n(&code[code_offset], n, buffer_data);
+        return n;
+    }
+
+    /// Selfdestruct the account (EVMC host method).
+    void selfdestruct(const address& addr, const address& beneficiary) noexcept override
+    {
+        record_account_access(addr);
+        recorded_selfdestructs.push_back({addr, beneficiary});
```
Comment (gumb0) [inline include/evmc/mocked_host.hpp]: could be `emplace_back`

File: include/evmc/mocked_host.hpp [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    /// Emit LOG (EVMC host method).
+    void emit_log(const address& addr,
+                  const uint8_t* data,
+                  size_t data_size,
+                  const bytes32 topics[],
+                  size_t topics_count) noexcept override
+    {
+        recorded_logs.push_back({addr, {data, data_size}, {}});
```
Comment (gumb0) [inline include/evmc/mocked_host.hpp]: `emplace_back`?

Comment (gumb0) [reply include/evmc/mocked_host.hpp]: Also probably you could just initialize topics vector inline with something like `std::vector<log_record>(&topics[0], topics + topics_count)`

File: include/evmc/mocked_host.hpp (hunk 3 above)
Comment (chfast) [reply include/evmc/mocked_host.hpp]: 🤷

File: include/evmc/mocked_host.hpp (hunk 4 above)
Comment (chfast) [reply include/evmc/mocked_host.hpp]: `emplace_back` does not work with `{}` — I believe it needs to know the type.

File: include/evmc/mocked_host.hpp (hunk 2 above)
Comment (chfast) [reply include/evmc/mocked_host.hpp]: I think it should. I missed the fact `std::vector` has `==`.

Comment (axic) [reply include/evmc/mocked_host.hpp]: Will you change this?

File: include/evmc/mocked_host.hpp [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+
+    /// The record of all block numbers for which get_block_hash() was called.
+    std::vector<int64_t> recorded_blockhashes;
+
+    /// The record of all account accesses.
+    std::vector<address> recorded_account_accesses;
+
+    /// The maximum number of entries in recorded_account_accesses record.
+    /// This is arbitrary value useful in fuzzing when we don't want the record to explode.
+    static constexpr auto max_recorded_account_accesses = 200;
```
Comment (axic) [inline include/evmc/mocked_host.hpp]: Why not power of 2?

File: include/evmc/mocked_host.hpp (hunk 1 above)
Comment (axic) [reply include/evmc/mocked_host.hpp]: Well 1283 is kind of irrelevant, only existed in Constantinople before Petersburg.

I know technically it is the same except the stipend, but still.

File: include/evmc/mocked_host.hpp [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+            recorded_calls.emplace_back(msg);
+            auto& call_msg = recorded_calls.back();
+            if (call_msg.input_size > 0)
+            {
+                m_recorded_calls_inputs.emplace_back(call_msg.input_data, call_msg.input_size);
+                const auto& input_copy = m_recorded_calls_inputs.back();
+                call_msg.input_data = input_copy.data();
+            }
+        }
+        return result{call_result};
```
Comment (axic) [inline include/evmc/mocked_host.hpp]: If there is a single `call_result` it is not possible to deal with precompiles? Well, this function can be overriden. Have a look at the precompile stuff soltest is doing.

File: include/evmc/mocked_host.hpp (hunk 2 above)
Comment (chfast) [reply include/evmc/mocked_host.hpp]: This is changed.

File: include/evmc/mocked_host.hpp (hunk 6 above)
Comment (chfast) [reply include/evmc/mocked_host.hpp]: Because this value is for humans.

File: include/evmc/mocked_host.hpp (hunk 7 above)
Comment (chfast) [reply include/evmc/mocked_host.hpp]: This PR is for moving MockedHost from evmone. I will handle solidity later.

File: include/evmc/mocked_host.hpp (hunk 2 above)
Comment (axic) [reply include/evmc/mocked_host.hpp]: I might be blind, but it still uses `std::equal` with iterators?

Comment (chfast) [reply include/evmc/mocked_host.hpp]: Yes, but these comments are attached to the outdated code fragment.

**Comment by chfast** [comment]
> I think it shouldn't refer to 1283 alone, but also to 2200.

Referencing EIP-2200 needs additional work to check its spec and it is still a draft.

**Pull request merged**